### PR TITLE
external: merge overlapping files concurrently

### LINKED ISSUE #48072: the performance of merge sort phase for global sort is poor
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

It use 1 thread to do merge sort, which is slow

### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #48072

Problem Summary:

### What is changed and how it works?
1. Merge files concurrently.
2. Adjust memory quota correctly.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [x] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/external/merge.go [hunk 1]
```diff
@@ -14,33 +20,86 @@ func MergeOverlappingFiles(
...
-	writerID string,
-	memSizeLimit uint64,
 	blockSize int,
 	writeBatchCount uint64,
 	propSizeDist uint64,
 	propKeysDist uint64,
 	onClose OnCloseFunc,
 ) error {
+	var dataFilesSlice [][]string
+	batchCount := len(paths) / int(variable.GetDDLReorgWorkerCounter())
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/merge.go:30]: this package is not only for DDL usage case.

**Review by ywqzzy — APPROVED** [review]

**Review by lance6716 — APPROVED** [review]

**Comment by D3Hunter** [comment]
```
=== RUN   TestGlobalSortMultiSchemaChange/ingest_dist_gs_backfill
[2023/11/01 16:57:23.842 +08:00] [INFO] [set.go:164] ["set global var"] [conn=4] [name=tidb_ddl_enable_fast_reorg] [val=1]
...
-- Test timed out at 2023-11-01 16:58:11 CST --
```

can this test run within 1m? or maybe enlarge timeout here:
```
    timeout = "short",
```

**Pull request merged**